File: NumberofIslands.py
class Solution(object):
    dx = [-1,1,0,0]
    dy = [0,0,1,-1]
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """

        islands = 0
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == '1':
                    self.explore(grid,i,j)
                    islands+=1
        return islands
        
    def explore(self, grid, x, y):
        grid[x][y] = 'X'
        for i in xrange(len(self.dx)):
            newx, newy = x+self.dx[i], y+self.dy[i]
            if 0 <= newx < len(grid) and 0<=newy < len(grid[0]) and grid[newx][newy] == '1':
                self.explore(grid, newx, newy)
        

File: numberOfIslands.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        val m = grid.size
        val n = grid[0].size

        fun bfs(a: Int, b: Int): Unit {
            grid[a][b] = '2'
            val dirs = setOf(0 to 1, -1 to 0, 1 to 0, 0 to -1)
            for((xa, xb) in dirs){
                val aa = a + xa
                val bb = b + xb
                if(aa >= 0 && aa < m && bb >= 0 && bb < n && grid[aa][bb] == '1'){
                    bfs(aa, bb)
                }
            }
        }
        var res = 0
        for(i in 0 until m){
            for(j in 0 until n){
                if(grid[i][j] == '1') {
                    bfs(i, j)
                    res ++
                }
            }
        }
        return res
    }

}
